### PR TITLE
Re-implemented dynamic loading of database drivers

### LINKED ISSUE #321: on-demand dynamic loading of DB drivers
Following the transition to Jbuilder/Dune, I can now implement that DB drivers are only loaded once we first try to establish a connection using them

### PR DESCRIPTION
**Pull request by frank-emrich**
Fixes #321.

The drivers are now only loaded when they are needed for establishing a connection.
One annoyance is that we still can't automatically load the dependencies of drivers, but need to do it manually. For example, when loading links-postgresql, we need to load postgresql first. In @dhil's hack these dependencies were hard-coded into the loader for each known driver. Now, each driver must ship a json file stating its dependencies. 

I've tested that it works with all 3 drivers.
Note that a simple "make" now doesn't build any drivers. I've added a new target "postgresql" to the Makefile to build the driver.

### CODE REVIEW

**Comment by dhil** [comment]
Thanks, I will look at it tomorrow.

File: links.opam [hunk 1]
```diff
@@ -64,4 +64,5 @@ depends: [
   "cohttp"
   "websocket-lwt"
   "safepass"
+  "ocamlfind"
```
Comment (dhil) [inline links.opam]: I think it is unnecessary to list `ocamlfind` as it is bundled with the compiler.

File: Makefile [hunk 2]
```diff
@@ -2,14 +2,22 @@
...
 
 native: nc
 
+postgresql:
+        #Re-state links here because otherwise the version of links installed in
+        #opam may be used, leading to a linking error when loading the driver into the
+        #version of links built in the current folder
+	jbuilder build -p links,links-postgresql -j 4 @install
+
+all: nc postgresql
```
[the remark this answers is not in the document]
Comment (frank-emrich) [reply Makefile]: But then we should at least create a new target that builds links without any drivers?

File: core/basicsettings.ml [hunk 3]
```diff
@@ -80,6 +80,21 @@ let config_file_path = match Utility.getenv "LINKS_CONFIG" with
           with End_of_file ->
             None
 
+
+let default_db_driver_search_folders =
+  let install_path = Filename.concat  (Sys.getcwd ()) "_build/install/default" in
```
[the remark this answers is not in the document]
Comment (dhil) [reply core/basicsettings.ml]: I am not sure I understand what you mean. If I understand your code then the build directory path is hard coded into the compiler.

[the remark this answers is not in the document]
Comment (frank-emrich) [reply core/basicsettings.ml]: No. I added a new setting called "db_driver_dirs". When attempting to load drivers, we currently first check the directories listed in this setting, then (if unsuccesful) try to use opam to find appropriate links-foodriver packages in our opam installation. The *default value* of the setting db_driver_dirs is such that it goes down to the _build subdirectory of the current directory where the drivers would end up if we build in the current directory. But if you provide your own value for the setting db_driver_dirs, you can override this behavior.

File: Makefile [hunk 4]
```diff
@@ -2,14 +2,22 @@
 .DEFAULT_GOAL: nc
 
 nc:
-	jbuilder build -p links,links-postgresql -j 4 @install
+	jbuilder build -p links -j 4 @install
 	@echo "#!/bin/sh" > links
-	@echo "LINKS_LIB=\"$(shell pwd)/_build/default/lib\" LINKS_LD_LIBRARY_PATH=\"$(shell pwd)/_build/default/database/pg-driver\" $(shell pwd)/_build/default/bin/links.exe \"\$$@\"" >> links
```
[the remark this answers is not in the document]
Comment (frank-emrich) [reply Makefile]: exactly, by setting the setting "db_driver_dirs" discussed in the other comments

File: core/basicsettings.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhil) [reply core/basicsettings.ml]: What is the syntax for the path string? Can I have multiple paths, e.g. `$HOME/foo/bar:$LIB/baz` etc?

[the remark this answers is not in the document]
Comment (dhil) [reply core/basicsettings.ml]: I think the settings name should be named `plugin_path` or something like that. It is potentially going to be more generic than just loading database drivers.

[the remark this answers is not in the document]
Comment (frank-emrich) [reply core/basicsettings.ml]: Yes, it uses the standard PATH syntax where : separates individual paths. Don't you think we should wait with renaming the setting until we have decided on how a more general dynamic loading system should look like?

File: core/jbuild [hunk 5]
```diff
@@ -10,5 +10,5 @@
   (wrapped false)
   (modes (native))
   (flags (:standard -safe-string -dtypes -w Ae-44-45-60 -g -cclib -lunix -thread))
-  (libraries (str yojson ppx_deriving_yojson.runtime unix safepass base64 ANSITerminal linenoise              cohttp lwt websocket websocket-lwt.cohttp))
+  (libraries (str yojson ppx_deriving_yojson.runtime unix safepass base64 ANSITerminal linenoise cohttp lwt websocket websocket-lwt.cohttp findlib))
```
[the remark this answers is not in the document]
Comment (frank-emrich) [reply core/jbuild]: But this is core/jbuild, where we need findlib because it is used to resolve opam package paths

File: links.opam (hunk 1 above)
Comment (frank-emrich) [reply links.opam]: Are you sure? You can use ocamlfind as a frontned for the ocaml compiler, but it doesn't seem to be included in it. In particular it's available via opam:
https://opam.ocaml.org/packages/ocamlfind/

File: core/jbuild (hunk 5 above)
[the remark this answers is not in the document]
Comment (dhil) [reply core/jbuild]: I am aware of that... But `findlib` is listed as a dependency in `bin/jbuild`.

File: core/basicsettings.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhil) [reply core/basicsettings.ml]: I think you may need to call `Utility.expand` (aka `Notfound.Sys.expand`) on each path array.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (frank-emrich) [reply Makefile]: But until now, nc would unconditionally try to build the postgres driver. Should we try to implement a check here that builds only the drivers whose dependencies are satisifed?

File: core/basicsettings.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (frank-emrich) [reply core/basicsettings.ml]: Fixed

File: core/jbuild (hunk 5 above)
[the remark this answers is not in the document]
Comment (frank-emrich) [reply core/jbuild]: Sorry, you're right. Fixed!

**Pull request merged**